### ISSUE TITLE
Extend it more

### ISSUE THREAD

**Issue by kaiokendev** [kaiokendev/cutoff-len-is-context-len#1]
you don't need to change max_position_embeddings here, you can instead add the variable to the model config.json, but you should set it to >2048 to get the correct sinusoidal
```
class LlamaRotaryEmbedding(torch.nn.Module):
    def __init__(self, dim, max_position_embeddings=2048, base=10000, device=None):
        super().__init__()
        max_position_embeddings = 8192
        inv_freq = 1.0 / (base ** (torch.arange(0, dim, 2).float().to(device) / dim))
        self.register_buffer("inv_freq", inv_freq)
       
        # Build here to make `torch.jit.trace` work.
        self.max_seq_len_cached = max_position_embeddings
        t = torch.arange(self.max_seq_len_cached, device=self.inv_freq.device, dtype=self.inv_freq.dtype)

        # Add these two lines of code, everything else remains the same
        scale = 1/4
        t *= scale
        ....
```

**Comment by kaiokendev** [comment]
@bojone Hello Su, am curious on your thoughts of the above approach, I explain it more here https://kaiokendev.github.io/til#extending-context-to-8k
Goal is to extend the context of the pre-trained model, only this approach worked for me (I found it at first by setting position IDs to repeat block wise within the bounds of pre-trained sequence length L, i.e. [1, 1, 1, 1, 2, 2, 2, 2, 3, 3, 3, 3, ....L, L, L, L] and found that it works well for extending the context during finetuning). Mainly asking because I tested a lot of approaches inspired by your blog approaches but only this one worked consistently, maybe you have ideas on how to make it better lol

**Comment by bojone** [comment]
Interesting discovery! However, some of my recent experiments have left me quite puzzled about this issue.

If you can read Chinese, please refer to https://kexue.fm/archives/9603. As I mentioned in the article, all mainstream length extrapolation schemes fail under the "Post Norm + GAU (a single-head attention)" combination, and only the HWFA proposed in the linked article is effective. I just tried your approach on it as well, and it also failed. In other words, the effectiveness of length extrapolation methods still depends on the model architecture, but obviously, mainstream research has not analyzed this point, which means our understanding of length extrapolation is far from sufficient.

**Comment by kaiokendev** [comment]
@bojone Yes, actually I read that article before but it was not leading to good extrapolation for me when applied to the pre-trained model. Like you said it seem architecture matters a lot. 😆 

I saw your HWFA before but mainly I was not satisfied with the point:
> 2、Full Attention加上RoPE，外推效果会下降；

But I also think they can be combined -- using the scaling only on layer L-1? I did not try it because it seemed modifying the layers on the pre-trained model was leading to poor results in the fine-tuning stage, and mainly the goal in my case is to extend the context of the pre-trained model (which seems to mirror length extrapolation)

I tried looking into the idea of "using interpolation to achieve extrapolation" but I could not find many papers. I'm also curious on your thoughts of the conditional positional encoding from Vision Transformers (ViT)? I did not see you wrote an article on it
https://arxiv.org/pdf/2102.10882.pdf
I am using machine translation, hope that is okay, but you write:
> 然而，尽管这样能确保平移不变性的出现，但是会带来另外的问题，最严重的就是由于每层的感受野被限制在w
内，注意力机制的能力大大削弱，导致训练效果不如常规注意力（下面称为Full Attention）。此外，我们对长度外推的期望其实不仅仅是“平移不变性”，而是“平移更好性”，也就是说越往后效果应该越好才对（比如In Context Learning场景，给的examples越多，效果应该越好），所以模型还应该要能捕捉全局依赖的能力。

The scaling factor seems similar to the CPE approach, as the position of `t` is dynamically adjusted based on input sequence length N to fit in the pre-trained length L. This would mimic the effect of translation invariance without reducing the receptive field to `w`. There is some tests currently to benchmark the quality https://github.com/ggerganov/llama.cpp/discussions/1965 . The hope would be to see the quality improves as N grows larger. I tried the `log n` scaling as well but the effect for extrapolation in fine-tuning was small, but I did not try it for interpolation yet.

I am also looking into ReLU, maybe you already tried it and it didn't improve anything: https://arxiv.org/pdf/2302.06461.pdf

**Comment by xdevfaheem** [comment]
> you don't need to change max_position_embeddings here, you can instead add the variable to the model config.json, but you should set it to >2048 to get the correct sinusoidal
> 
> ```
> class LlamaRotaryEmbedding(torch.nn.Module):
>     def __init__(self, dim, max_position_embeddings=2048, base=10000, device=None):
>         super().__init__()
>         max_position_embeddings = 8192
>         inv_freq = 1.0 / (base ** (torch.arange(0, dim, 2).float().to(device) / dim))
>         self.register_buffer("inv_freq", inv_freq)
>        
>         # Build here to make `torch.jit.trace` work.
>         self.max_seq_len_cached = max_position_embeddings
>         t = torch.arange(self.max_seq_len_cached, device=self.inv_freq.device, dtype=self.inv_freq.dtype)
> 
>         # Add these two lines of code, everything else remains the same
>         scale = 1/4
>         t *= scale
>         ....
> ```

I wonder what if shrink the position more. i mean, what if pos 1 looks like pos .125 and so on. also, can we use this directly when inferencing?

**Comment by bojone** [comment]
@kaiokendev Today, I tried your approach on Falcon-7b, but it didn't work. I set the scale to 1/2 and then input a sequence of about 4k tokens, with the output being almost like gibberish.

A more effective improvement is to divide the sequence into three parts: Context + Prompt + Output. For the Context part, evenly divide the range [0, 1024] as position ids, while for the Prompt + Output, use the normal position ids 1024, 1025, 1025, etc. This method is somewhat similar to the Parallel Context Window.

However, the above approaches are not as effective as NBCE. You can compare them with NBCE ( https://github.com/bojone/NBCE ).

**Comment by kaiokendev** [comment]
@bojone That is weird, it should achieve similar result to ALiBi, as per the below chart:

![248038968-0b08e754-0f01-4a33-85f8-876c16bee68a](https://github.com/kaiokendev/cutoff-len-is-context-len/assets/129691954/ab95acb6-1f8a-432e-8941-9b473203027a)

I read the NBCE but my understanding is it mainly for in-context learning and cannot work for e.g. long range summarization, is that wrong understanding?

Edit: I shoud note the effect is mainly seen from finetuning with the change. Still, it should not be gibberish...

**Comment by bojone** [comment]
> I read the NBCE but my understanding is it mainly for in-context learning and cannot work for e.g. long range summarization, is that wrong understanding?

For long text summarization, NBCE may not perform well, but for long text question-answering that doesn't rely heavily on the order of contexts, it should yield good results.

**Comment by kaiokendev** [comment]
For reference, here is the specific code I am using: https://github.com/kaiokendev/cutoff-len-is-context-len/blob/main/util/dope_llama_monkey_patch.py

> For long text summarization, NBCE may not perform well, but for long text question-answering that doesn't rely heavily on the order of contexts, it should yield good results.

Yes I observed that, but the problem is we would like to achieve the proper length extrapolation, i.e. long range summarization outside the domain of question answering, which the scale seems to perform well on which is why I'm surprised you got gibberish.

**Comment by afoland** [comment]
Have you tried anything along the lines of t *= exp(-t/5909)  ?  This would keep low-numbered token positionings the same as originally trained, and modify significantly only at larger lengths.  You'd expect that the models can't learn a whole lot in detail about the absolute positioning at large position count numbers, but the early ones could definitely have strong learnable patterns with absolute position.

(5909 is chosen to give a factor of 4 by the time you hit 8192, so that 8192 looks like 2048).

**Comment by kaiokendev** [comment]
@afoland I am not sure if I am doing it wrong? This is the plot I am seeing from that operation
![Figure_1](https://github.com/kaiokendev/cutoff-len-is-context-len/assets/129691954/f2265425-8feb-4671-964a-1acb59a85d6a)
But this will not work properly because it the positions are repeated and reversing
But of the general idea of some sort of logarithmic position scaling, I have not tried it yet

**Comment by afoland** [comment]
You're right, I didn't think about it carefully enough, the pre-existing t gives it a maximum.  Back to the drawing board.  But even the plot above shows a usable context expansion by roughly a factor e for context lengths up to about 5560 (before it turns around.)  So the optimal version of this form of equation is to use 5567 in the exponential, expanding 2048 to 5567 smoothly.

The intent is to find a function that:
*Maps low numbers to nearly themselves
*Maps large numbers to a selectable factor 1/k times themselves (the exponential scaling is not selectable, it maxes out at e)
*Is continuous
*Is monotonic between 1 and k x 2048

**Comment by kaiokendev** [comment]
@bojone Hello Su, it seems Meta has concurrently validated the approach. Here is the paper if you are interested: https://arxiv.org/abs/2306.15595 "Extending Context Window of Large Language Models via Positional Interpolation"

**Comment by bojone** [comment]
> @bojone Hello Su, it seems Meta has concurrently validated the approach. Here is the paper if you are interested: https://arxiv.org/abs/2306.15595 "Extending Context Window of Large Language Models via Positional Interpolation"

Thanks for sharing, it's an amazing progress. But it seems that the meta approach still needs additional fine-tuning. If I only modify the freq's scale, I do encounter gibberish or repetition here —— the generated result is not natural text at all.

**Comment by kaiokendev** [comment]
@bojone Yes, it was the same case in my own version. I did not find a way to achieve long range extrapolation without fine-tuning, otherwise the result is gibberish, but after fine-tuning it approaches parity with extended pre-trained length.

**Comment by bojone** [comment]
> @bojone Yes, it was the same case in my own version. I did not find a way to achieve long range extrapolation without fine-tuning, otherwise the result is gibberish, but after fine-tuning it approaches parity with extended pre-trained length.

I see, I thought it was a method that didn't require fine-tuning, sorry, it was my misunderstanding.

Since fine-tuning is needed, how long is the fine-tuning length? Is it the original 2048? Or is it longer like 8192?

**Comment by kaiokendev** [comment]
The fine-tune in my case is 4096, however it is reported the extrapolation is still gained even without fine-tuning, but the cost is increased perplexity. With fine-tuning, the perplexity is recovered completely. With only 300 samples of 4096 it is possible to learn the interpolated positions. I provide a perplexity measure here:

![ppl_mine](https://github.com/kaiokendev/cutoff-len-is-context-len/assets/129691954/4acd253f-37d5-4735-91f3-53739016bba1)

In the case of the Meta paper, only 0.5% of pre-training corpus is needed for 32k context length.

**Comment by bojone** [comment]
> The fine-tune in my case is 4096, however it is reported the extrapolation is still gained even without fine-tuning, but the cost is increased perplexity. With fine-tuning, the perplexity is recovered completely. With only 300 samples of 4096 it is possible to learn the interpolated positions. I provide a perplexity measure here:
> 
> ![ppl_mine](https://user-images.githubusercontent.com/129691954/249339286-4acd253f-37d5-4735-91f3-53739016bba1.png)
> 
> In the case of the Meta paper, only 0.5% of pre-training corpus is needed for 32k context length.

Thank you, I understand now. That is to say, in order to handle a text of length L, I still need to fine-tune with a text of length L, but after adding this scaling technique, the number of samples required for fine-tuning will be much less.

**Comment by kaiokendev** [comment]
@bojone Yes, although, even without fine-tuning it should not produce gibberish in certain cases. Here, `base` is not fine-tuned, but after applying scaling, the perplexity is still OK for scale of `0.5` and L of `4092` even though pre-trained L is `2048`

**Comment by kaiokendev** [comment]
@bojone I think there was a concern because the approach needed finetuning, whereas NBCE does not require finetuning? Someone has made an improvement to the method, with Neural Tangent Kernel Aware Interpolation. By modifying the base instead of the position scale, the results are even better without fine-tuning, as the features are not disturbed.
```
old_init = transformers.models.llama.modeling_llama.LlamaRotaryEmbedding.__init__
def ntk_scaled_init(self, dim, max_position_embeddings=2048, base=10000, device=None):

    #The method is just these three lines
    max_position_embeddings = 16384
    a = 8 #Alpha value
    base = base * a ** (dim / (dim-2)) #Base change formula

    old_init(self, dim, max_position_embeddings, base, device)
```

I wonder if this approach is more intuitive? It also does not require finetuning, so I wonder what result you will get this time ^^

Source: https://www.reddit.com/r/LocalLLaMA/comments/14lz7j5/ntkaware_scaled_rope_allows_llama_models_to_have/

**Comment by bojone** [comment]
@kaiokendev This is an exciting result, I'm sorry I just saw the information, I'm going to understand and try it now.

However, regardless of the result, I still think that NBCE has its unique value, because it seems that NTK's solution can only extend llama's context length to about 8k, and I have tested that even under 50k context, NBCE can still output the correct results.

**Comment by bojone** [comment]
@kaiokendev This is so astonishing! I just verified it on the GAU+RoPE+Post Norm model, and it works indeed. This is the second solution that can improve extrapolation effects without fine-tuning under this combination! (The first one is the architecture-agnostic NBCE)

Next, I will take a good look at "Neural Tangent Kernel Aware Interpolation", I am ashamed that I have not heard of such a technique before.

**Comment by kaiokendev** [comment]
> NTK's solution can only extend llama's context length to about 8k

Ah, actually it seems it can extend a lot, I think with fine-tuning it can be more.

According to Meta's paper, the scaling interpolation has a limit of 1.2 million tokens with fine-tuning, but according to the author of NTK, as the features are not disturbed, it should be able to go to billion token range as the fidelity is preserved, but this does not account for the key-value limitation or the saturation of attention score, but I think the "theoretical limit" is quite high. It would need further study to show what the true limit is without fine-tuning. For example, I did not try it with log-n scaling

Author of NTK approach is @bloc97

**Comment by bojone** [comment]
> > NTK's solution can only extend llama's context length to about 8k
> 
> Ah, actually it seems it can extend a lot, I think with fine-tuning it can be more.
> 
> According to Meta's paper, the scaling interpolation has a limit of 1.2 million tokens with fine-tuning, but according to the author of NTK, as the features are not disturbed, it should be able to go to billion token range as the fidelity is preserved, but this does not account for the key-value limitation or the saturation of attention score, but I think the "theoretical limit" is quite high. It would need further study to show what the true limit is without fine-tuning. For example, I did not try it with log-n scaling
> 
> Author of NTK approach is @bloc97

In the graph provided by the author, the larger the alpha, the better the extrapolation effect, but the ppl will also gradually rise.

Fine-tuning will certainly improve, but as you said, I am more concerned about the limit without fine-tuning.

**Comment by bojone** [comment]
I tested it on the Chinese fine-tuned llama-7b, and it really works! Kudos to the author again! But overall, under the same case, its effect is not as good as that of nbce.

**Comment by StrangeTcy** [comment]
> @bojone Yes, it was the same case in my own version. I did not find a way to achieve long range extrapolation without fine-tuning, otherwise the result is gibberish, but after fine-tuning it approaches parity with extended pre-trained length.

I can anecdotally confirm that
